From ada9316ad6e55c9e5986c0b7bcf74355bdc4f813 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <grinberg@compulab.co.il>
Date: Wed, 27 Feb 2013 19:06:56 +0200
Subject: ARM: i.MX6: cm-fx6: add support for SPI bus 1

Add support for SPI bus 1 and peripheral DataImage SCF0403 LCD
connected to the bus.

Signed-off-by: Igor Grinberg <grinberg@compulab.co.il>
---
 arch/arm/mach-mx6/board-cm-fx6-dl.h |    6 ++++++
 arch/arm/mach-mx6/board-cm-fx6-q.h  |    6 ++++++
 arch/arm/mach-mx6/board-cm-fx6.c    |   33 +++++++++++++++++++++++++++++++++
 3 files changed, 45 insertions(+)

diff --git a/arch/arm/mach-mx6/board-cm-fx6-dl.h b/arch/arm/mach-mx6/board-cm-fx6-dl.h
index 58ad714..e06608f 100644
--- a/arch/arm/mach-mx6/board-cm-fx6-dl.h
+++ b/arch/arm/mach-mx6/board-cm-fx6-dl.h
@@ -79,6 +79,12 @@ static iomux_v3_cfg_t cm_fx6_dl_pads[] = {
 	MX6DL_PAD_EIM_D18__ECSPI1_MOSI,
 	MX6DL_PAD_EIM_EB2__GPIO_2_30,	/* SS0 */
 	MX6DL_PAD_EIM_D19__GPIO_3_19,	/* SS1 */
+	/* eCSPI2 */
+	MX6DL_PAD_EIM_CS0__ECSPI2_SCLK,
+	MX6DL_PAD_EIM_OE__ECSPI2_MISO,
+	MX6DL_PAD_EIM_CS1__ECSPI2_MOSI,
+	MX6DL_PAD_EIM_D24__GPIO_3_24,	/* SS2 */
+	MX6DL_PAD_EIM_D25__GPIO_3_25,	/* SS3 */
 
 	/* I2C1 */
 	MX6DL_PAD_EIM_D21__I2C1_SCL,
diff --git a/arch/arm/mach-mx6/board-cm-fx6-q.h b/arch/arm/mach-mx6/board-cm-fx6-q.h
index 0ab9823..6ca3b55 100644
--- a/arch/arm/mach-mx6/board-cm-fx6-q.h
+++ b/arch/arm/mach-mx6/board-cm-fx6-q.h
@@ -79,6 +79,12 @@ static iomux_v3_cfg_t cm_fx6_q_common_pads[] = {
 	MX6Q_PAD_EIM_D18__ECSPI1_MOSI,
 	MX6Q_PAD_EIM_EB2__GPIO_2_30,	/* SS0 */
 	MX6Q_PAD_EIM_D19__GPIO_3_19,	/* SS1 */
+	/* eCSPI2 */
+	MX6Q_PAD_EIM_CS0__ECSPI2_SCLK,
+	MX6Q_PAD_EIM_OE__ECSPI2_MISO,
+	MX6Q_PAD_EIM_CS1__ECSPI2_MOSI,
+	MX6Q_PAD_EIM_D24__GPIO_3_24,	/* SS2 */
+	MX6Q_PAD_EIM_D25__GPIO_3_25,	/* SS3 */
 
 	/* I2C1 */
 	MX6Q_PAD_EIM_D21__I2C1_SCL,
diff --git a/arch/arm/mach-mx6/board-cm-fx6.c b/arch/arm/mach-mx6/board-cm-fx6.c
index 802b04b..aad3c5b 100644
--- a/arch/arm/mach-mx6/board-cm-fx6.c
+++ b/arch/arm/mach-mx6/board-cm-fx6.c
@@ -32,6 +32,7 @@
 #include <linux/spi/spi.h>
 #include <linux/spi/flash.h>
 #include <linux/spi/ads7846.h>
+#include <linux/spi/scf0403.h>
 #include <linux/i2c.h>
 #include <linux/i2c/pca953x.h>
 #include <linux/i2c/at24.h>
@@ -84,10 +85,13 @@
 #define CM_FX6_ECSPI1_CS0		IMX_GPIO_NR(2, 30)
 #define CM_FX6_GREEN_LED		IMX_GPIO_NR(2, 31)
 #define CM_FX6_ECSPI1_CS1		IMX_GPIO_NR(3, 19)
+#define CM_FX6_ECSPI2_CS2		IMX_GPIO_NR(3, 24)
+#define CM_FX6_ECSPI2_CS3		IMX_GPIO_NR(3, 25)
 #define CM_FX6_USB_HUB_RST		IMX_GPIO_NR(7, 8)
 
 #define SB_FX6_SD3_WP			IMX_GPIO_NR(7, 0)
 #define SB_FX6_SD3_CD			IMX_GPIO_NR(7, 1)
+#define SB_FX6_LCD_RST			IMX_GPIO_NR(8, 11)
 
 #define MX6_ARM2_LDB_BACKLIGHT		IMX_GPIO_NR(1, 9)
 #define MX6_ARM2_USB_OTG_PWR		IMX_GPIO_NR(3, 22)
@@ -300,11 +304,21 @@ static int cm_fx6_spi0_cs[] = {
 	CM_FX6_ECSPI1_CS1,	/* CS1 */
 };
 
+static int cm_fx6_spi1_cs[] = {
+	CM_FX6_ECSPI2_CS2,	/* CS0 */
+	CM_FX6_ECSPI2_CS3,	/* CS1 */
+};
+
 static const struct spi_imx_master cm_fx6_spi0_data = {
 	.chipselect	= cm_fx6_spi0_cs,
 	.num_chipselect	= ARRAY_SIZE(cm_fx6_spi0_cs),
 };
 
+static const struct spi_imx_master cm_fx6_spi1_data = {
+	.chipselect	= cm_fx6_spi1_cs,
+	.num_chipselect	= ARRAY_SIZE(cm_fx6_spi1_cs),
+};
+
 #if defined(CONFIG_MTD_M25P80) || defined(CONFIG_MTD_M25P80_MODULE)
 static struct mtd_partition cm_fx6_spi_flash_partitions[] = {
 	{
@@ -384,6 +398,10 @@ static void __init ads7846_init(void)
 static inline void ads7846_init(void) {}
 #endif /* CONFIG_TOUCHSCREEN_ADS7846 */
 
+static struct scf0403_pdata scf0403_config = {
+	.reset_gpio	= SB_FX6_LCD_RST,
+};
+
 static struct spi_board_info cm_fx6_spi0_board_info[] __initdata = {
 #if defined(CONFIG_MTD_M25P80) || defined(CONFIG_MTD_M25P80_MODULE)
 	{
@@ -407,6 +425,18 @@ static struct spi_board_info cm_fx6_spi0_board_info[] __initdata = {
 #endif
 };
 
+static struct spi_board_info cm_fx6_spi1_board_info[] __initdata = {
+#if defined(CONFIG_LCD_SCF0403) || defined(CONFIG_LCD_SCF0403_MODULE)
+	{
+		.modalias               = "scf0403",
+		.max_speed_hz           = 1000000,
+		.bus_num                = 1,
+		.chip_select            = 1,
+		.platform_data          = &scf0403_config,
+	},
+#endif
+};
+
 static void __init spi_register_bus_binfo(int busnum,
 					  const struct spi_imx_master *spidata,
 					  struct spi_board_info *info,
@@ -432,6 +462,9 @@ static void __init cm_fx6_spi_init(void)
 	spi_register_bus_binfo(0, &cm_fx6_spi0_data,
 			       cm_fx6_spi0_board_info,
 			       ARRAY_SIZE(cm_fx6_spi0_board_info));
+	spi_register_bus_binfo(1, &cm_fx6_spi1_data,
+			       cm_fx6_spi1_board_info,
+			       ARRAY_SIZE(cm_fx6_spi1_board_info));
 }
 #else /* CONFIG_SPI_IMX */
 static inline void cm_fx6_spi_init(void) {}
-- 
1.7.9.5

